File: erettsegi/emelt/2024_may/python/Belepteto.py
from dataclasses import dataclass
from datetime import time, timedelta

@dataclass
class Mozgas:
    azonosito: str
    idopont: time
    tipus: int


with open('bedat.txt') as bedat:
    mozgast_keszit = lambda split: Mozgas(split[0], time.fromisoformat(split[1]), int(split[2]))
    mozgasok = [ mozgast_keszit(k.split(' ')) for k in bedat.readlines() ]

print(f'2. Feladat: Első belépő: {mozgasok[0].idopont}, utolsó kilépő: {mozgasok[-1].idopont}')

keso_kezdet = time(7, 50)
keso_zaras = time(8, 15)
kesok = (f'{k.idopont} {k.azonosito}' for k in mozgasok if k.idopont > keso_kezdet and k.idopont < keso_zaras)

with open('kesok.txt', 'w') as output: output.writelines(kesok)

menzasok_szama = sum(1 for k in mozgasok if k.tipus == 3)

print(f'4. Feladat: Menzások száma: {menzasok_szama}')

kolcsonzesek_szama = len(set(k.azonosito for k in mozgasok if k.tipus == 4))

print(f'5. Feladat: Könyvtári kölcsönzések száma: {kolcsonzesek_szama}.\n{"Többen voltak" if kolcsonzesek_szama > menzasok_szama else "Nem voltak többen"}, mint a menzán')

szunet_kezdete = time(10, 45)
kapu_zaras = time(10, 50)
szunet_vege = time(11, 0)

szunetbe_ismert_kilepok = [ k.azonosito for k in mozgasok if k.idopont > szunet_kezdete and k.idopont < szunet_vege and k.tipus == 2 ]
szunet_elott_mar_belepok = [ k.azonosito for k in mozgasok if k.idopont < szunet_kezdete ]
szunet_vegeig_belepok = (k.azonosito for k in mozgasok if k.idopont > kapu_zaras and k.idopont < szunet_vege and k.tipus == 1 and \
                                                          not k.azonosito in szunetbe_ismert_kilepok and k.azonosito in szunet_elott_mar_belepok)

print(f'6. Feladat: {str.join(" ", szunet_vegeig_belepok)}')

bekert_azonosito = input('7. Feladat: Kérem egy tanuló azonosítóját! ')
bekert_tanulo_mozgasai = [ k for k in mozgasok if k.azonosito == bekert_azonosito ]
utolso_ido = bekert_tanulo_mozgasai[-1].idopont
elso_ido = bekert_tanulo_mozgasai[1].idopont
eltelt_ido_mp = (timedelta(hours = utolso_ido.hour, minutes = utolso_ido.minute) - timedelta(hours = elso_ido.hour, minutes = elso_ido.minute)).total_seconds()

print(f'Eltelt idő: {int(eltelt_ido_mp / 3600)} óra {int(eltelt_ido_mp / 60) % 60} perc')

File: erettsegi/emelt/2024_may/rust/Belepteto.rs
use std::{fs::File, io::{stdin, BufRead, BufReader, Write}};
use chrono::NaiveTime;
use itertools::Itertools;

fn main() {
    let mozgasok = BufReader::new(File::open("bedat.txt").unwrap())
                            .lines()
                            .map(|k| {
                                let line = k.unwrap();
                                let split = line.split(' ').into_iter().collect::<Vec<_>>();

                                Mozgas { azonosito: split[0].to_string(), idopont: split[1].parse().unwrap(), tipus: split[2].parse().unwrap()}
                            })
                            .collect::<Vec<_>>();

    println!("2. Feladat: Első belépő: {}, utolsó kilépő: {} ", mozgasok[0].idopont, mozgasok[mozgasok.len() - 1].idopont);


    let keso_kezdet = NaiveTime::from_hms_opt(7, 50, 0).unwrap();
    let keso_zaras = NaiveTime::from_hms_opt(8, 15, 0).unwrap();

    let kesok_fileba = mozgasok.iter()
                               .filter(|k| k.idopont > keso_kezdet && k.idopont < keso_zaras)
                               .map(|k| format!("{} {}", k.idopont, k.azonosito))
                               .join("\n");

    write!(File::create("kesok.txt").unwrap(), "{}", kesok_fileba);

    let menzasok_szama = mozgasok.iter()
                                 .filter(|k| k.tipus == 3)
                                 .count();

    println!("4. Feladat: Menzások száma: {menzasok_szama}");

    let kolcsonzesek_szama = mozgasok.iter()
                                     .filter(|k| k.tipus == 4)
                                     .map(|k| &k.azonosito)
                                     .unique()
                                     .count();

    println!("5. Feladat: Könyvtári kölcsönzések száma: {kolcsonzesek_szama}\n{}, mint a menzán", if kolcsonzesek_szama > menzasok_szama { "Többen voltak" } else { "Nem voltak többen" });

    let szunet_kezdete = NaiveTime::from_hms_opt(10, 45, 0).unwrap();
    let kapu_zaras = NaiveTime::from_hms_opt(10, 50, 0).unwrap();
    let szunet_vege = NaiveTime::from_hms_opt(11, 0, 0).unwrap();

    let szunetbe_ismert_kilepok = mozgasok.iter()
                                          .filter(|k| k.idopont > szunet_kezdete && k.idopont < szunet_vege)
                                          .filter(|k| k.tipus == 2)
                                          .map(|k| &k.azonosito)
                                          .collect::<Vec<_>>();

    let szunet_elott_mar_belepok = mozgasok.iter()
                                           .filter(|k| k.idopont < szunet_kezdete)
                                           .map(|k| &k.azonosito)
                                           .collect::<Vec<_>>();

    let szunet_vegeig_belepok = mozgasok.iter()
                                        .filter(|k| k.idopont > kapu_zaras && k.idopont < szunet_vege)
                                        .filter(|k| k.tipus == 1)
                                        .map(|k| &k.azonosito)
                                        .filter(|k| !szunetbe_ismert_kilepok.contains(k) && szunet_elott_mar_belepok.contains(k))
                                        .join(" ");

    println!("6. Feladat: {szunet_vegeig_belepok}");
    println!("7. Feladat: Kérem egy tanuló azonosítóját!");

    let mut bekert_azonosito_str = String::new();
    stdin().read_line(&mut bekert_azonosito_str);

    let bekert_azonosito = bekert_azonosito_str.trim_end();
    let bekert_tanulo_mozgasai = mozgasok.iter()
                                         .filter(|k| k.azonosito == bekert_azonosito)
                                         .collect::<Vec<_>>();

    let eltelt_ido = bekert_tanulo_mozgasai[bekert_tanulo_mozgasai.len() - 1].idopont - bekert_tanulo_mozgasai[0].idopont;

    println!("Eltelt idő: {} óra {} perc", eltelt_ido.num_hours(), eltelt_ido.num_minutes() - (eltelt_ido.num_hours() * 60));
}

struct Mozgas {
    azonosito: String,
    idopont: NaiveTime,
    tipus: i32
}